xonsh/environ.py

```diff
@@ -82,21 +92,15 @@ def format_prompt(template=default_prompt):
     + NO_COLOR -- Resets any previously used color codes
     """
     env = builtins.__xonsh_env__
-    cwd = env['PWD']
-    branch = current_branch(cwd=cwd)
-    branch = '' if branch is None else ' ' + branch
-    p = template.format(
-            user=env.get('USER', '<user>'),
-            hostname=socket.gethostname(),
-            cwd=cwd.replace(env['HOME'], '~'),
-            curr_branch=branch,
-            **TERM_COLORS
-            )
-    return p
+    fmt = env.get('FORMAT_DICT', FORMAT_DICT)
+    fmt = {k:v for (k,v) in fmt.items() if '{{{}}}'.format(k) in template}
```

Comment:
Right you are.  I usually use `%` in my own code, so I didn't even know that anything fancy was possible with `format`.  I think we can't get this right with a simple regex, but we could use `string.Formatter.parse` to do the right thing without any hassle (the names will be the second element in each tuple it returns).